### ISSUE TITLE
Unhandled HTTP 500 "String index of out of range" error

### ISSUE THREAD

**Issue by amoeba** [NCEAS/metacat#1495]
I typo'd a Metacat URL and ended up with an unhandled except that really oughta be handled with a 400.

To reproduce:

1. Go to https://knb.ecoinformatics.org/knb/metacat/d1/mn/v2/meta/foo
2.  See HTTP Status 500 – Internal Server Error: `java.lang.StringIndexOutOfBoundsException: String index out of range: -1`

You can see that the URL I meant to type was https://knb.ecoinformatics.org/knb/d1/mn/v2/meta/foo (w/o 'metacat' after 'knb').

**Comment by mbjones** [comment]
Agreed. Good catch. Any non-existent service paths should return 404 rather than 500. @taojing2002 can you patch this up when convenient in an upcoming release? Sounds like it needs some better exception handling in the dispatch methods. Probably another case where we're catching `Exception` rather than dealing with different exception types properly.

**Comment by amoeba** [comment]
Thanks for taking a look, @taojing2002. That explanation makes sense.

However, I think that it's still a bug here. I realize this is subjective, but I think it's an ultimate goal that public-facing HTTP endpoints to never 500 as 500s are for exceptional and unhandled behavior. Parsing and handling an HTTP request shouldn't really result in exceptional behavior. I think we can produce a 400-series code here, whatever's appropriate. Probably 404.

I'd be happy to take a look if you'd like. Let me know.
